src/tls/TLSClient.fst

```diff
@@ -0,0 +1,298 @@
+  let ad: StatefulPlain.adata id = StatefulPlain.makeAD id ct in
+  // Range.frange -> Range.range
+  let rg: Range.frange id = 0, length text in
+  // DataStream.fragment -> DataStream.pre_fragment -> bytes
+  let f: DataStream.fragment id rg = text |> unsafe_coerce in
+  // LHAEPlain.plain -> StatefulPlain.plain -> Content.fragment
+  //NS: Not sure about the unsafe_coerce: but, it's presence clearly means that #id cannot be inferred
+  let f: LHAEPlain.plain id ad rg = Content.CT_Data #id rg f |> unsafe_coerce in
+  // StatefulLHAE.cipher -> StatefulPlain.cipher -> bytes
+  // FIXME: without the three additional #-arguments below, extraction crashes
+  StatefulLHAE.encrypt #id w ad rg f
+
+let decryptRecord_TLS12_AES_GCM_128_SHA256 rd ct cipher = 
+  let ad: StatefulPlain.adata id = StatefulPlain.makeAD id ct in
+  let (Some d) = StatefulLHAE.decrypt #id rd ad cipher in
+  Content.repr id d
+
+let hsbuf = alloc ([] <: list (hs_msg * bytes))
+
+type callbacks = FFICallbacks.callbacks
+
+val sendTcpPacket: callbacks:callbacks -> buf:bytes -> (result (bytes))
+let sendTcpPacket callbacks buf = 
+  let result = FFICallbacks.ocaml_send_tcp callbacks (get_cbytes buf) in
+  if result < 0 then
+    Error (AD_internal_error, "socket send failure")
+  else
+    Correct buf
+
+val recvTcpPacket: callbacks:callbacks -> maxlen:Prims.int -> (result (bytes))    
+let recvTcpPacket callbacks maxlen =
+  let str = String.make maxlen 0z in
+  let recvlen = FFICallbacks.ocaml_recv_tcp callbacks str  in
+  if recvlen < 0 then
+    Error (AD_internal_error, "socket receive failure")
+  else
+    let b = String.substring str 0 recvlen in
+    let ab = abytes b in
+    Correct ab
+  
+private val really_read_rec_callback: b:bytes -> callbacks -> l:nat -> (result (lbytes (l+length b)))
+
+let rec really_read_rec_callback prev callbacks len = 
+    if len = 0 
+    then Correct prev
+    else 
+      match recvTcpPacket callbacks len with
+      | Correct b -> 
+            let lb = length b in
+      	    if lb = len then Correct(prev @| b)
+      	    else really_read_rec_callback (prev @| b) callbacks (len - lb)
+      | Error e -> Error e
+
+private let really_read_callback = really_read_rec_callback empty_bytes
+
+val recvRecordCallback: callbacks -> protocolVersion -> 
+  (result (Content.contentType * protocolVersion * b:bytes { length b <= max_TLSCiphertext_fragment_length}))
+let recvRecordCallback callbacks pv =
+  match really_read_callback callbacks 5 with 
+  | Correct header -> (
+      match Record.parseHeader header with  
+      | Correct (ct,pv,len) -> (
+         match really_read_callback callbacks len with
+         | Correct payload  -> Correct (ct,pv,payload)
+         | Error e          -> Error e )
+      | Error e             -> Error e )
+  | Error e                 -> Error e
+
+(*let hsbuf = alloc ([] <: list (hs_msg * bytes)) *)
+
+let recvHSRecordCallback12 callbacks pv kex log = 
+  let (hs_msg, to_log) = match !hsbuf with
```

Comment:
(note: not sure if this is code you wrote or if these were the parts that were in the test client already... if that was already written, feel free to dismiss)